include/sharpSAT/primitive_types.h

```diff
@@ -8,20 +8,189 @@
 #ifndef SHARP_SAT_PRIMITIVE_TYPES_H_
 #define SHARP_SAT_PRIMITIVE_TYPES_H_
 
+#include <ostream>
+#include <limits>
+#include <cassert>
+
 namespace sharpSAT {
 
-typedef unsigned VariableIndex;
+struct VariableIndex {
+  unsigned var_id_;
+
+public:
+  VariableIndex() : var_id_(0) {}
+  explicit VariableIndex(unsigned var_id) : var_id_(var_id) {}
+
+  explicit operator unsigned() const { return var_id_; }
+
+  bool operator ==(const VariableIndex& rhs) const
+  { return var_id_ == rhs.var_id_; }
+
+  bool operator !=(const VariableIndex& rhs) const
+  { return var_id_ != rhs.var_id_; }
+
+  bool operator <(const VariableIndex& rhs) const
+  { return var_id_ < rhs.var_id_; }
+
+  bool operator >(const VariableIndex& rhs) const
+  { return var_id_ > rhs.var_id_; }
+
+  bool operator <=(const VariableIndex& rhs) const
+  { return var_id_ <= rhs.var_id_; }
+
+  bool operator >=(const VariableIndex& rhs) const
+  { return var_id_ >= rhs.var_id_; }
+
+  VariableIndex& operator ++() {
+    ++var_id_;
+    return *this;
+  }
+
+  VariableIndex operator ++(int) {
+    auto before = *this;
+    ++var_id_;
+    return before;
+  }
+
+  VariableIndex& operator --() {
+    --var_id_;
+    return *this;
+  }
+
+  VariableIndex operator --(int) {
+    auto before = *this;
+    --var_id_;
+    return before;
+  }
+};
+
+inline std::ostream& operator <<(std::ostream& os, const VariableIndex var) {
+  return os << static_cast<unsigned>(var);
+}
+
+static const VariableIndex varsSENTINEL(0);
```

Comment:
~This variable should be `extern` and defined in only one TU.~

Nevermind, that would be costly in terms of performance. However, we should note that the _identity_ (address) of these variables is not guaranteed across different TUs and cannot be relied upon.

Or go with C++17 and use inline variables.